File: p55.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool canJump(vector<int>& nums) {
        int i = 0;
        int end_index = nums[0];
        while(i <= end_index && i < nums.size()) {
            end_index = max(end_index, nums[i]+i);
            ++i;
        }
        return end_index >= nums.size()-1;
    }
};

File: p55.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        furthest = 0

        for i, n in enumerate(nums):
            if i <= furthest:
                furthest = max(furthest, i+n)
            else:
                return False

        return furthest >= n-1
